Support running task pods in other Kubernetes namespaces (#17738)
apache/druid: docs/development/extensions-core/k8s-jobs.md

## Configuration

To use this extension please make sure to [include](../../configuration/extensions.md#loading-extensions) `druid-kubernetes-overlord-extensions` in the extensions load list for your overlord process.

The extension uses `druid.indexer.runner.capacity` to limit the number of k8s jobs in flight. A good initial value for this would be the sum of the total task slots of all the middle managers you were running before switching to K8s based ingestion. The K8s task runner uses one thread per Job that is created, so setting this number too large can cause memory issues on the overlord. Additionally set the variable `druid.indexer.runner.namespace` to the namespace in which you are running druid.

Other configurations required are:
`druid.indexer.runner.type: k8s` and `druid.indexer.task.encapsulatedTask: true`

### Running Task Pods in Another Namespace

> It is possible to run task pods in a different namespace from the rest of your Druid cluster.

Reviewer: since this feature is for pretty advanced kubernetes users and some of the complexity seems to be around trying to support the K8sTaskAdapter based pod adapters, would it make sense to just require using the pod template adapter if you want to use this feature?

Reviewer 2: That makes sense  because for non pod template mode, tasks are scheduled in the same ns with overlord

Author: done

If you are running multiple Druid clusters and would like to have a dedicated namespace for all your task pods, you can make the following changes to the runtime properties for your Overlord deployment:

- `druid.indexer.runner.namespace`: The namespace where the task pods will run.

> - `druid.indexer.runner.overlordNamespace`: The namespace where the Overlord resides.

Reviewer 2: ```suggestion
- `druid.indexer.runner.overlordNamespace`: The namespace where the Overlord resides. This must be defined when tasks are scheduled in different namespace.
```

Druid will tag Kubernetes jobs with a `druid.overlord.namespace` label. This label helps Druid filter out Kubernetes jobs belonging to other namespaces. When `druid.indexer.runner.overlordNamespace` is not provided, Druid will assume that the entire cluster operates within a single namespace and will default to `druid.indexer.runner.namespace`.

#### Dealing with ZooKeeper Problems

> Ensure that when you are running task pods in another namespace, your task pods are able to communicate with ZooKeeper from the overlord namespace. If you are using custom pod templates as described below, you can configure `druid.zk.service.host` to your ZooKeeper service.

Reviewer 2: ```suggestion
Ensure that when you are running task pods in another namespace, your task pods are able to communicate with ZooKeeper which might be deployed in the same namespace with overlord. If you are using custom pod templates as described below, you can configure `druid.zk.service.host` to your tasks.
```

> In all the above cases, Druid will match the selector to any value of task type. Druid applies similar logic for `dataSource`. For `context.tags` setting `null` or an empty object `{}` is equivalent. 

Reviewer 2: what are these changes about?

Author: Not sure why it is highlighted in red. Doing `git diff` shows no changes done in this line (as well as the Properties header3) below.
